include/seastar/core/abort_source.hh

```diff
@@ -193,6 +194,68 @@ public:
     }
 };
 
+class abort_on_either {
+    seastar::abort_source* _as0 = nullptr;
+    seastar::abort_source* _as1 = nullptr;
+    seastar::abort_source _as;
+    optimized_optional<seastar::abort_source::subscription> _sub0;
+    optimized_optional<seastar::abort_source::subscription> _sub1;
+
+private:
+    void do_subscribe() noexcept {
+        auto callback = [this] (const std::optional<std::exception_ptr>& opt_ex) noexcept {
+            if (!_as.abort_requested()) {
+                _as.do_request_abort(std::move(opt_ex));
+            }
+        };
+        _sub0 = _as0->subscribe(callback);
+        if (!_sub0) {
+            _as.do_request_abort(_as0->get_exception());
+        } else {
+            _sub1 = _as1->subscribe(callback);
+            if (!_sub1) {
+                _as.do_request_abort(_as1->get_exception());
+                _sub0 = {};
+            }
+        }
+    }
+public:
+    abort_on_either() = default;
+    abort_on_either(seastar::abort_source& as0, seastar::abort_source& as1) noexcept
```

Comment:
Why not make it more generic with `seastar::abort_source&... as`

Reply:
I took a stab at it, and it seemed overly complicated.
This probably needs better c++ meta-programming skills than mine.

Reply:
done in v4